### PR TITLE
add events to check passwords with the password policy app

### PR DESCRIPTION
**Pull request by schiessle**
Send "validate password events" to allow the password policy app (https://github.com/nextcloud/password_policy) to validate the password.

### CODE REVIEW

**Comment by LukasReschke** [comment]
Any need of a new hook over `'\OC\Share', 'verifyPassword'`? 

I'm fine with adding this one here as public API as well. But then the old one should die, it's non-public API anyways.

**Comment by schiessle** [comment]
> Any need of a new hook over '\OC\Share', 'verifyPassword'?

I would like to have this new event as some kind of "public API" that other apps can check their passwords as well. This already speaks against a hook in the OC namespace. Further the symphony events have some advantages. E.g. the password checker can throw a exception and you can handle it outside of the hook, while the old hook system swallows all exception.

Let's try to have clean implementations of new stuff.

> I'm fine with adding this one here as public API as well. But then the old one should die, it's non-public API anyways.

We can do this in a second step.

**Comment by schiessle** [comment]
Ready to review. You need the password policy app to test it: https://github.com/nextcloud/password_policy/pull/1

Password policy should be checked for public links and for the database backend password change in the user management and password change in the personal settings

cc @LukasReschke @MorrisJobke @blizzz

**Comment by LukasReschke** [comment]
@schiessle Merge conflict 🙈

File: settings/ChangePassword/Controller.php [hunk 1]
```diff
@@ -39,16 +41,20 @@ public static function changePersonalPassword($args) {
...
-		} else {
-			\OC_JSON::error();
+		try {
+			if (!is_null($password) && \OC_User::setPassword($username, $password)) {
+				\OC_JSON::success(['data' => ['message' => $l->t('Saved')]]);
+			} else {
+				\OC_JSON::error();
+			}
+		} catch (HintException $e) {
+			return \OC_JSON::error(['data' => ['message' => $e->getHint()]]);
```
Comment (LukasReschke) [inline settings/ChangePassword/Controller.php]: That's a void method. No return required.

File: settings/ChangePassword/Controller.php [hunk 2]
```diff
@@ -39,16 +41,20 @@ public static function changePersonalPassword($args) {
...
 			\OC_JSON::error(array("data" => array("message" => $l->t("Wrong password")) ));
 			exit();
 		}
-		if (!is_null($password) && \OC_User::setPassword($username, $password)) {
-			\OC_JSON::success();
-		} else {
-			\OC_JSON::error();
+		try {
+			if (!is_null($password) && \OC_User::setPassword($username, $password)) {
+				\OC_JSON::success(['data' => ['message' => $l->t('Saved')]]);
```
Comment (LukasReschke) [inline settings/ChangePassword/Controller.php]: While touching also move to the modern `OCP\JSON` approach.

**Comment by schiessle** [comment]
@LukasReschke merge conflict resolved.

File: settings/ChangePassword/Controller.php (hunk 2 above)
Comment (schiessle) [inline settings/ChangePassword/Controller.php]: Modern?? `\OCP\JSON` is already deprecated :see_no_evil:

**Comment by LukasReschke** [comment]
LGTM 👍

**Comment by MorrisJobke** [comment]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Backport? cc @karlitschek

**Comment by karlitschek** [comment]
please backport 👍

**Comment by schiessle** [comment]
backport for 9.0.52 https://github.com/nextcloud/server/pull/240

**Pull request merged**